src/sliding_windows.rs

```diff
@@ -0,0 +1,183 @@
+use std::cell::{Cell, UnsafeCell};
+use std::ops::{Deref, DerefMut};
+use std::fmt;
+
+/// This holds the backing allocation for the `Window` of a `SlidingWindowAdaptor`.
+/// 
+/// See [`.sliding_windows()`](trait.Itertools.html#method.sliding_windows) for more information.
+pub struct SlidingWindowStorage<T> {
+    window_size: usize,
+    /// acts as a refcount
+    uniquely_owned: Cell<bool>,
+    data: UnsafeCell<Vec<T>>
+}
+
+/* FIXME: uncomment this once it gets stable
+impl !Send for SlidingWindowStorage {}
+impl !Sync for SlidingWindowStorage {}
+*/
+
+impl<T> SlidingWindowStorage<T> {
+    /// Create a new `SlidingWindowStorage` with a given window size.
+    /// This will allocate as much memory as is needed to store the Window automatically.
+    ///
+    /// See [`.sliding_windows()`](trait.Itertools.html#method.sliding_windows) for more information.
+    pub fn new(window_size: usize) -> SlidingWindowStorage<T> {
+        SlidingWindowStorage {
+            window_size: window_size,
+            uniquely_owned: Cell::new(true),
+            data: UnsafeCell::new(Vec::with_capacity(window_size))
+        }
+    }
+
+    fn new_window<'a>(&'a self) -> Window<'a, T> {
+        // assert that the last window went out of scope
+        assert!(self.uniquely_owned.get(), "next() called before previous Window went out of scope");
+
+        self.uniquely_owned.set(false);
+
+        Window { drop_flag: &self.uniquely_owned, data: &self.data }
+    }
+
+    // push value onto self
+    fn push(&self, elt: T) -> bool {
+        assert!(self.uniquely_owned.get(), "next() called before previous Window went out of scope");
+        let data = unsafe { &mut *self.data.get() };
+        if data.len() != self.window_size {
+            data.push(elt);
+        } else {
+            data.remove(0);
```

Comment:
Asymptotically yes. Vec will be better when windows are small. I guess we'll have to take a deque to not be catastrophic on larger sizes yeah. It can be fixed later.